pmlc/dialect/tile/lowering.cc

```diff
@@ -355,12 +425,64 @@ struct AffineDomainOpConversion : public LoweringBase {
     rewriter.create<stripe::TerminateOp>(op->getLoc());
     rewriter.replaceOp(op, {outputTensor});
 
-    // TODO: use_default
     // TODO: no_defract
-    // TODO: NeedsInitialize
 
     return matchSuccess();
   }
+
+  bool needsInitialize(            //
+      Value* outRef,               //
+      stripe::TensorType outType,  //
+      ArrayRef<int64_t> ranges,    //
+      ArrayRef<Value*> constraints) const {
+    // Check if we have a simple output: 1 unique index per dimension, each full range
+    // If not, presume we need initialization for safety
+    // We assume here that the 0'th refinement is the output refinement
+    std::unordered_set<mlir::BlockArgument*> outIdxs;
+    auto refineOp = llvm::cast<stripe::RefineOp>(outRef->getDefiningOp());
+    const auto& outShape = outType.getShape();
+    for (unsigned i = 0; i < outShape.size(); i++) {
+      stripe::AffinePolynomial poly{refineOp.getOffset(i)};
+      if (poly == stripe::AffinePolynomial() && outShape[i].size == 1) {
+        continue;
+      }
+      if (poly.constant || poly.terms.size() != 1 || poly.terms.begin()->second != 1) {
+        // Not a single index with a multiplier of 1, bail
+        return true;
+      }
+      // TODO: should we check all terms?
+      auto idx = poly.terms.begin()->first;
+      if (outIdxs.count(idx)) {
+        // The index isn't unique, bail
+        return true;
+      }
+      outIdxs.insert(idx);
+      auto range = ranges[idx->getArgNumber()];
+      if (range != outShape[i].size) {
+        // The index range does not cover the entire size of the tensor dimension
+        return true;
+      }
+    }
+
+    // Now we check if we have any constraints that are 'output only'
+    // Output only indexes actually reduce the range we write to, whereas constraints
+    // that use both input + output make writes but only process some of the input
```

Comment:
Technically this isn't true (as I recently realized when looking into the old version of this code while trying to figure out how to do padding).  But apparently, it's a sufficient approximation is every case we ever hit, so good enough for now!